common/src/api-types.ts

```diff
@@ -0,0 +1,122 @@
+import * as t from 'io-ts'
+
+type Field<
+  Base,
+  T extends Base | undefined,
+  K extends keyof Base,
+  Fallback = undefined
+> = T extends {
+  [k in K]: Base[K]
+}
+  ? T[K]
+  : Fallback
+type TypeOf<T> = T extends t.Any ? t.TypeOf<T> : {}
+
+export namespace Schema {
+  export interface Request {
+    readonly parameters?: t.Any
+    readonly query?: t.Any
+    readonly body?: t.Any
+  }
+
+  export interface Response {
+    readonly body?: t.Any
+  }
+
+  export interface Method {
+    readonly request?: Request
+    readonly response?: Response
+  }
+
+  export type MethodName = 'all' | 'get' | 'post' | 'put' | 'delete' | 'patch' | 'options' | 'head'
+  export type Methods = Readonly<Partial<Record<MethodName, Method>>>
+
+  export type Path = `/${string}`
+  export type RootPath = '/'
+  export type Children = Readonly<Partial<Record<Exclude<Path, RootPath>, Endpoint>>>
+
+  export interface Endpoint {
+    readonly methods?: Methods
+    readonly children?: Children
+  }
+  export type API = Endpoint
+}
+
+export namespace Get {
+  export type MethodName<Endpoint extends Schema.Endpoint> = Schema.MethodName &
+    (Endpoint extends { methods: Schema.Methods } ? keyof Endpoint['methods'] : never)
+
+  export function methodNames<Endpoint extends Schema.Endpoint>(
+    endpoint: Endpoint
+  ): Set<Get.MethodName<Endpoint>> {
+    const methodNames = new Set<MethodName<Endpoint>>()
+    if (endpoint.methods === undefined) return methodNames
+    for (const key in endpoint.methods) {
+      // Typescript isn't smart enough to figure this out on its own
+      methodNames.add(key as MethodName<Endpoint>)
+    }
+    return methodNames
+  }
+
+  export type Path<Endpoint extends Schema.Endpoint> = Schema.Path &
+    (Endpoint extends { methods: Schema.Methods }
+      ? keyof Endpoint['children'] | Schema.RootPath
```

Comment:
👀 nice catch, that's supposed to be an `&`

Reply:
So this is the weird thing about `Endpoint`. The `.methods` field corresponds to the methods available at the `/` endpoint. The `.children` field corresponds to the endpoints for the children. It's messy, but the `/` endpoint is fundamentally different from all the others in the sense that it can't have any further children.

`Schema.RootPath` is `/`, so this just means "if the endpoint has a `.methods`, then it supports the `/` path as well as all of its children"